### ISSUE TITLE
[question] embedding yt-dlp in Pyodide terminal.. library is available but network requests get `TransportError('timed out')`

### ISSUE THREAD

**Issue by warren-bank** [yt-dlp/yt-dlp#11957]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm asking a question and **not** reporting a bug or requesting a feature
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar questions **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Please make sure the question is worded well enough to be understood

[Here](https://warren-bank.github.io/single-page-apps/yt-dlp-web-console/index.html) is a work-in-progress. It's pretty much entirely boilerplate.
The 1st demo script works, and displays the version of `yt-dlp`.
The 2nd demo is a verbatim example of [embedding `yt-dlp`](https://github.com/yt-dlp/yt-dlp#embedding-yt-dlp) from your README: [Extracting information](https://github.com/yt-dlp/yt-dlp#extracting-information)

<details>
  <summary>Here is the output:</summary>

  ```text
    [youtube] Extracting URL: https://www.youtube.com/watch?v=qhKci8jY510
    [youtube] qhKci8jY510: Downloading webpage
    WARNING: [youtube] Unable to download webpage: timed out
    [youtube] qhKci8jY510: Downloading ios player API JSON
    WARNING: [youtube] timed out. Retrying (1/3)...
    [youtube] qhKci8jY510: Downloading ios player API JSON
    WARNING: [youtube] timed out. Retrying (2/3)...
    [youtube] qhKci8jY510: Downloading ios player API JSON
    WARNING: [youtube] timed out. Retrying (3/3)...
    [youtube] qhKci8jY510: Downloading ios player API JSON
    WARNING: [youtube] Unable to download API page: timed out (caused by TransportError('timed out'))
    [youtube] qhKci8jY510: Downloading iframe API JS
    WARNING: [youtube] Unable to download webpage: timed out
    [youtube] qhKci8jY510: Downloading mweb player API JSON
    WARNING: [youtube] timed out. Retrying (1/3)...
    [youtube] qhKci8jY510: Downloading mweb player API JSON
    WARNING: [youtube] timed out. Retrying (2/3)...
    [youtube] qhKci8jY510: Downloading mweb player API JSON
    WARNING: [youtube] timed out. Retrying (3/3)...
    [youtube] qhKci8jY510: Downloading mweb player API JSON
    WARNING: [youtube] Unable to download API page: timed out (caused by TransportError('timed out'))

    ERROR: [youtube] qhKci8jY510: Failed to extract any player response; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U

    Traceback (most recent call last):
      File "/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 1769, in __extract_info
        ie_result = ie.extract(url)
                    ^^^^^^^^^^^^^^^
      File "/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 742, in extract
        ie_result = self._real_extract(url)
                    ^^^^^^^^^^^^^^^^^^^^^^^
      File "/lib/python3.11/site-packages/yt_dlp/extractor/youtube.py", line 4435, in _real_extract
        webpage, master_ytcfg, player_responses, player_url = self._download_player_responses(url, smuggled_data, video_id, webpage_url)
                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/lib/python3.11/site-packages/yt_dlp/extractor/youtube.py", line 4399, in _download_player_responses
        player_responses, player_url = self._extract_player_responses(
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/lib/python3.11/site-packages/yt_dlp/extractor/youtube.py", line 4042, in _extract_player_responses
        raise ExtractorError('Failed to extract any player response')

    yt_dlp.utils.ExtractorError: [youtube] qhKci8jY510: Failed to extract any player response; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U

    During handling of the above exception, another exception occurred:
    Traceback (most recent call last):
      File "<console>", line 2, in <module>
      File "/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 1623, in extract_info
        return self.__extract_info(url, self.get_info_extractor(key), download, extra_info, process)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 1652, in wrapper
        self.report_error(str(e), e.format_traceback())
      File "/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 1092, in report_error
        self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
      File "/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 1031, in trouble
        raise DownloadError(message, exc_info)

    yt_dlp.utils.DownloadError: ERROR: [youtube] qhKci8jY510: Failed to extract any player response; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  ```
</details>

<details>
  <summary>I've attempted to trace the code to find the networking http client:</summary>

  ```text
    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/YoutubeDL.py#L1769

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/extractor/common.py#L742

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/extractor/youtube.py#L4435
    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/extractor/youtube.py#L4399
    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/extractor/youtube.py#L3916
    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/extractor/youtube.py#L4042
  ```

  - - - -

  ```text
    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/extractor/youtube.py#L787
      _download_ytcfg() calls: self._download_webpage()

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/extractor/common.py#L6
      import http.client

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/extractor/common.py#L916
      def _download_webpage_handle(
        self, url_or_request, video_id, note=None, errnote=None, fatal=True,
        encoding=None, data=None, headers={}, query={}, expected_status=None,
        impersonate=None, require_impersonation=False
      )

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/extractor/common.py#L1176
      def _download_webpage(
        self, url_or_request, video_id, note=None, errnote=None,
        fatal=True, tries=1, timeout=NO_DEFAULT, *args, **kwargs
      )

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/extractor/common.py#L842
      def _request_webpage(
        self, url_or_request, video_id, note=None, errnote=None, fatal=True, data=None,
        headers=None, query=None, expected_status=None, impersonate=None, require_impersonation=False
      )
    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/extractor/common.py#L898
        return self._downloader.urlopen(self._create_request(url_or_request, data, headers, query, extensions))

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/extractor/common.py#L783
      def set_downloader(self, downloader):
        """Sets a YoutubeDL instance as the downloader for this IE."""
        self._downloader = downloader

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/YoutubeDL.py#L4151
      def urlopen(self, req):
        """ Start an HTTP download """
    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/YoutubeDL.py#L4172
        return self._request_director.send(req)

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/YoutubeDL.py#L4251
      def _request_director(self):
        return self.build_request_director(_REQUEST_HANDLERS.values(), _RH_PREFERENCES)

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/YoutubeDL.py#L4215
      def build_request_director(self, handlers, preferences=None):
        director = RequestDirector(logger=logger, verbose=self.params.get('debug_printtraffic'))

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/networking/common.py#L51
      class RequestDirector:
    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/networking/common.py#L117
        response = handler.send(request)

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/networking/common.py#L149
      class RequestHandler(abc.ABC):

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/networking/_requests.py#L256
      class RequestsRH(RequestHandler, InstanceStoreMixin):
        """Requests RequestHandler
        https://github.com/psf/requests
        """
        _SUPPORTED_URL_SCHEMES = ('http', 'https')

    https://github.com/yt-dlp/yt-dlp/blob/2024.12.23/yt_dlp/networking/_requests.py#L315
      def _send(self, request):
  ```
</details>

...but I don't see any obvious problems.

I'm hoping that somebody here who knows this repo's code far better than I ever could might offer some helpful insight.. as to why network requests are failing when the library is used in this execution context. I rarely touch Python, so any help with this would be greatly appreciated.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

command 1:
```python
yt_dlp.version.__version__
```

output 1:
```text
'2024.12.23'
```

command 2:
```python
URL = 'https://www.youtube.com/watch?v=qhKci8jY510'

ydl_opts = {"verbose": True}

with yt_dlp.YoutubeDL(ydl_opts) as ydl:
    info = ydl.extract_info(URL, download=False)
    print(json.dumps(ydl.sanitize_info(info)))
```

output 2:
```text
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out missing (_WriteStream) (No ANSI), error missing (_WriteStream) (No ANSI), screen missing (_WriteStream) (No AN
SI)
[debug] yt-dlp version stable@2024.12.23 from yt-dlp/yt-dlp [65cf46cdd] (pip) API
[debug] params: {'verbose': True, 'compat_opts': set(), 'http_headers': {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) 
Chrome/91.0.4472.124 Safari/537.36', 'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8', 'Accept-Language': 'en-us,en;q=0.5', 'Sec-Fetch-Mode': 
'navigate'}}
[debug] Python 3.11.2 (CPython wasm32 32bit) - Emscripten-3.1.32-wasm32-32bit (OpenSSL 1.1.1n  15 Mar 2022, glibc 2.14)
[debug] exe versions: none
[debug] Optional libraries: none
[debug] Proxy map: {}
[debug] Request Handlers: urllib
[debug] Loaded 1837 extractors
[youtube] Extracting URL: https://www.youtube.com/watch?v=qhKci8jY510
[youtube] qhKci8jY510: Downloading webpage
WARNING: [youtube] Unable to download webpage: [Errno 23] Host is unreachable
[youtube] qhKci8jY510: Downloading ios player API JSON
WARNING: [youtube] [Errno 23] Host is unreachable. Retrying (1/3)...
[youtube] qhKci8jY510: Downloading ios player API JSON
WARNING: [youtube] [Errno 23] Host is unreachable. Retrying (2/3)...
[youtube] qhKci8jY510: Downloading ios player API JSON
WARNING: [youtube] [Errno 23] Host is unreachable. Retrying (3/3)...
[youtube] qhKci8jY510: Downloading ios player API JSON
WARNING: [youtube] Unable to download API page: [Errno 23] Host is unreachable (caused by TransportError('[Errno 23] Host is unreachable'))
[youtube] qhKci8jY510: Downloading iframe API JS
WARNING: [youtube] Unable to download webpage: [Errno 23] Host is unreachable
[youtube] qhKci8jY510: Downloading mweb player API JSON
WARNING: [youtube] [Errno 23] Host is unreachable. Retrying (1/3)...
[youtube] qhKci8jY510: Downloading mweb player API JSON
WARNING: [youtube] [Errno 23] Host is unreachable. Retrying (2/3)...
[youtube] qhKci8jY510: Downloading mweb player API JSON
WARNING: [youtube] [Errno 23] Host is unreachable. Retrying (3/3)...
[youtube] qhKci8jY510: Downloading mweb player API JSON
WARNING: [youtube] Unable to download API page: [Errno 23] Host is unreachable (caused by TransportError('[Errno 23] Host is unreachable'))
ERROR: [youtube] qhKci8jY510: Failed to extract any player response; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriat
e issue template. Confirm you are on the latest version using  yt-dlp -U
  File "/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 742, in extract
    ie_result = self._real_extract(url)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/yt_dlp/extractor/youtube.py", line 4435, in _real_extract
    webpage, master_ytcfg, player_responses, player_url = self._download_player_responses(url, smuggled_data, video_id, webpage_url)
                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/yt_dlp/extractor/youtube.py", line 4399, in _download_player_responses
    player_responses, player_url = self._extract_player_responses(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/yt_dlp/extractor/youtube.py", line 4042, in _extract_player_responses
    raise ExtractorError('Failed to extract any player response')
Traceback (most recent call last):
  File "/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 1769, in __extract_info
    ie_result = ie.extract(url)
                ^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 742, in extract
    ie_result = self._real_extract(url)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/yt_dlp/extractor/youtube.py", line 4435, in _real_extract
    webpage, master_ytcfg, player_responses, player_url = self._download_player_responses(url, smuggled_data, video_id, webpage_url)
                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/yt_dlp/extractor/youtube.py", line 4399, in _download_player_responses
    player_responses, player_url = self._extract_player_responses(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/yt_dlp/extractor/youtube.py", line 4042, in _extract_player_responses
    raise ExtractorError('Failed to extract any player response')
yt_dlp.utils.ExtractorError: [youtube] qhKci8jY510: Failed to extract any player response; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , fil
ling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "<console>", line 2, in <module>
  File "/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 1623, in extract_info
    return self.__extract_info(url, self.get_info_extractor(key), download, extra_info, process)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 1652, in wrapper
    self.report_error(str(e), e.format_traceback())
  File "/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 1092, in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
  File "/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 1031, in trouble
    raise DownloadError(message, exc_info)
yt_dlp.utils.DownloadError: ERROR: [youtube] qhKci8jY510: Failed to extract any player response; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q=
 , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
```

**Comment by bashonly** [comment]
It's an issue with your network/DNS

**Comment by warren-bank** [comment]
Totally agree that it's a networking issue..
and that the culprit is in the debug log: `[Errno 23] Host is unreachable`

I can reach youtube.com in a regular browser window..
so it must have something to do with Pyodide.

I'll need to google for similar (Pyodide related) issues..
I completely realize this question/request is off-topic,
nevertheless, I was hoping somebody might know the answer off the top of their head..
if there's no such solution after a few days, I'll gladly close this issue.

**Comment by bashonly** [comment]
It's likely CORS-related?

**Comment by bashonly** [comment]
see https://github.com/yt-dlp/yt-dlp/issues/7940#issuecomment-1918162141 and the discussion that follows

**Comment by warren-bank** [comment]
thanks! I haven't read it all yet, but it does seem relevant. I'll take a closer look now.

**Comment by warren-bank** [comment]
* https://warren-bank.github.io/single-page-apps/yt-dlp-web-console/index.html
  ```text
    Mixed Content: The page at '<URL>' was loaded over HTTPS, but attempted to connect to the insecure WebSocket endpoint 'ws://<URL>'. This request has been blocked; this endpoint must be available over WSS.
  ```
* http://localhost/index.html
  ```text
    WebSocket connection to 'ws://<URL>' failed: WebSocket is closed before the connection is established.
  ```

- - - -

so.. you're right about this issue having to do with WebSockets.
Huge thanks for pointing me in that direction!

Follow-up question..
are WebSockets required to use all of the [webpage extractors](https://github.com/ytdl-org/youtube-dl/tree/master/youtube_dl/extractor).. or only Youtube?

**Comment by warren-bank** [comment]
failed tests:
* [The CW](https://www.cwtv.com/shows/superman-and-lois/the-end-and-the-beginning/?play=b333019d-39b0-473c-91d6-bfe87c503690)
* [CBS](https://www.cbs.com/shows/video/zheaHAj4F4XuO1lltJR0s0m_abRtifsU/)

**Comment by bashonly** [comment]
yt-dlp itself doesn't really use websockets (only for niconico and twitcasting IIRC)

I think maybe you had the wrong takeaway from the discussion I linked. The issue is CORS, which causes *any* requests will be blocked

From https://github.com/yt-dlp/yt-dlp/issues/9820#issuecomment-2166712378 (also linked in the discussion I linked to earlier):

> With a bit of monkey patching, I was able to get yt-dlp running in Pyodide and download some videos.
Not that it is very useful, as I needed to disable all CORS security features. I also didn't bother adding ffmpeg support. This would probably require translating the CLI arguments yt-dlp uses to the ffmpeg.wasm api. Threading also doesn't seem to work.

> The main advantage to running yt-dlp on a user's device, compared to on a server, is that no one needs to pay for the server's bandwidth and the traffic is coming from the user's IP and not some datacenter IP. Using a CORS proxy would negate that.
> If you have the user install some browser extension to bypass the CORS limitations, you might as well ask them to install a program. And if you have them install an Electron App, why not ship the yt-dlp executable instead of trying to get yt-dlp running in wasm.

**Comment by warren-bank** [comment]
I'm not very concerned about CORS, because ultimately.. my intention is to package this into a userscript that only acts when a web page is visited that has a supported _yt-dlp_ extractor. Presumably, requests made by the _yt-dlp_ extractor would be allowed by CORS when made from the same origin.

I'll need to test a bit more if the reason WebSockets are failing is the result of CORS, despite there not being any log messages to indicate that this is the case. If so, then I could probably work around it. If not, and it has something to do with how Pyodide performs networking.. then idk?

**Comment by seproDev** [comment]
I am out of the country right now, so can't look at my code from back then. Iirc. the main thing was just patching some failing imports to no longer be fatal. I actually upstreamed some of those changes, so that shouldn't be a problem anymore.
Other than that I don't remember any changes to yt-dlp being required.

**Comment by warren-bank** [comment]
@seproDev ..huge thanks for a response :smiley: ..much appreciated!

**Comment by warren-bank** [comment]
fwiw, a few additional notes..

I added the following line to my `hosts` file:
```text
127.0.0.1 youtube.com
```

..and then tested:
* http://youtube.com/index.html
  - as an alias for: http://localhost/index.html
  - console log:
    ```text
      Firefox can’t establish a connection to the server at ws://www.youtube.com:443/.
        GET wss://www.youtube.com/
      Firefox can’t establish a connection to the server at ws://www.youtube.com:443/.
      The connection to ws://www.youtube.com:443/ was interrupted while the page was loading.
    ```
  - where: `www.youtube.com` still resolves to a Youtube server

**Comment by warren-bank** [comment]
I think that I [found](https://github.com/koenvo/pyodide-http/issues/33#issuecomment-1880700871) the magical incantation..

tldr;
```python
  import pyodide_http; pyodide_http.patch_all()
```

#### tests:

command 1:
```python
URL = 'https://www.youtube.com/watch?v=qhKci8jY510'

ydl_opts = {"debug_printtraffic": True}

with yt_dlp.YoutubeDL(ydl_opts) as ydl:
    info = ydl.extract_info(URL, download=False)
    print(json.dumps(ydl.sanitize_info(info)))
```

output 1 (abridged):
```text
[youtube] qhKci8jY510: Downloading webpage
director: Handler preferences for this request: urllib=0
director: Checking if "urllib" supports this request.
director: Sending request via "urllib"
```

- - - -

command 2:
```python
import urllib
urllib.request.urlopen('https://www.youtube.com/watch?v=qhKci8jY510')
```

output 2 (Firefox console log):
```text
[error] Firefox can’t establish a connection to the server at ws://www.youtube.com:443/.
```

- - - -

command 3:
```python
import urllib
urllib.request.urlopen('https://warren-bank.github.io/single-page-apps/yt-dlp-web-console/index.html')

# note: server response includes header: "access-control-allow-origin: *"
```

output 3 (Firefox console log):
```text
[error] Firefox can’t establish a connection to the server at ws://warren-bank.github.io:443/.
```

- - - -

command 4:
```javascript
await pyodide.loadPackage('pyodide_http');
```

```python
import pyodide_http; pyodide_http.patch_all()

import urllib
urllib.request.urlopen('https://warren-bank.github.io/single-page-apps/yt-dlp-web-console/index.html')

# note: server response includes header: "access-control-allow-origin: *"
```

output 4:
```text
<http.client.HTTPResponse object at 0x1d9a160>
```

output 4 (Firefox console log):
```text
[warning] Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help http://xhr.spec.whatwg.org/
```

- - - -

command 5:
```javascript
await pyodide.loadPackage('pyodide_http');
```

```python
import pyodide_http; pyodide_http.patch_all()

import urllib
urllib.request.urlopen('https://www.youtube.com/watch?v=qhKci8jY510')
```

output 5:
```text
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/lib/python3.12/site-packages/pyodide_http/_urllib.py", line 31, in urlopen
    resp = send(request)
           ^^^^^^^^^^^^^
  File "/lib/python3.12/site-packages/pyodide_http/_core.py", line 121, in send
    xhr.send(to_js(request.body))
pyodide.ffi.JsException: NetworkError: A network error occurred.
```

output 5 (Firefox console log):
```text
[warning] Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help http://xhr.spec.whatwg.org/
[error] Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://www.youtube.com/watch?v=qhKci8jY510. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200.
```

**Comment by warren-bank** [comment]
I'll close this issue. The solution in my previous comment is simple and effective.

Sorry again. I completely realize (now) that this question was off-topic, and only pertains to the Pyodide runtime environment.
I really appreciate everyone's helpful attitude.
